src/ty_check/constraint_check.rs

```diff
@@ -0,0 +1,75 @@
+use crate::ast::{IdentKinded, WhereClauseItem};
+
+use crate::ty_check::unify::{ConstrainMap, Constrainable};
+
+use super::unify::substitute;
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ConstraintScheme {
+    pub generics: Vec<IdentKinded>,
+    pub implican: Vec<WhereClauseItem>,
+    pub implied: WhereClauseItem
+}
+
+#[derive(Debug, Clone)]
+pub struct ConstraintEnv {
+    theta: Vec<ConstraintScheme>
+}
+
+impl ConstraintScheme {
+    pub fn new(implied: &WhereClauseItem) -> Self {
+        ConstraintScheme { generics: vec![], implican: vec![], implied: implied.clone() }
+    }
+
+    pub fn is_where_clause_item(&self) -> bool {
+        self.generics.len() == 0 && self.implican.len() == 0
+    }
+}
+
+impl ConstraintEnv {
+    pub fn new() -> Self {
+        Self { theta: Vec::new() }
+    } 
+
+    pub fn append_constraint(&mut self, con: &ConstraintScheme) {
+        self.theta.push(con.clone());
+    }
+
+    pub fn append_constraints(&mut self, cons: &Vec<ConstraintScheme>) {
+        self.theta.extend(cons.clone());
+    }
+
+    pub fn remove_constraints(&mut self, cons: &Vec<ConstraintScheme>) {
+        cons.iter().for_each(|con_remove| {
+            self.theta
+            .swap_remove(
+                self.theta
+                .iter()
+                .position(|con|
+                    con == con_remove).unwrap()
+                );
+        });
+    }
+
+    pub fn check_predicate(&self, goal: &WhereClauseItem) -> bool {
+        let mut constr_map = ConstrainMap::new();
+        let mut prv_rels = Vec::new();
+
+        self.theta.iter().find(|con| {
+            constr_map.clear();
+            prv_rels.clear();
+            
+            let mut goal_clone = goal.clone();
+
+            if goal_clone.constrain(&mut con.implied.clone(), &mut constr_map, &mut prv_rels).is_ok() {
+                con.implican.iter().fold(true, |res, c| {
+                    let mut goal = c.clone();
+                    substitute(&constr_map, &mut goal);
+                    res && self.check_predicate(&goal)
+                })
+            } else {
+                false
+            }
+        }).is_some()
+    }
```

Comment:
Diesen Algorithmus müsste ich evtl. auch noch mal ändern. Im Allgemeinen terminiert er nicht z.B. 
```impl<T> MyTrait for T where T: MyTrait { ... }``` was zu dem ConstraintScheme $$\forall T. T:MyTrait => T:MyTrait$$ korrespondiert. Wenn man da so was wie i32:MyTrait prüfen will, kommt man in eine Endlosschleife, da dann die einzige Regel angewendet wird und anschließend die Vorraussetzung i32:MyTrait geprüft wird...

Reply:
occurs_check nutzen
bind_to_ident

Reply:
Ist gelöst.
Statt so einer rekursiven Version wird jetzt eine iterative Version mit Backtracking genutzt. Dabei findet immer, wenn ein goal durch mehrere subgoals ersetzt wird,  ein occurs-check statt. Dieser überprüft, ob eines dieser subgoals auf dem Backtracking-Stack schon vorkommt. Ist dies der Fall, wurde ein Zyklus entdeckt und es findet keine Ersetzung des aktuallen goals durch die subgoals statt.